lib/src/validator/strict_dependencies.dart

```diff
@@ -2,49 +2,76 @@
 // for details. All rights reserved. Use of this source code is governed by a
 // BSD-style license that can be found in the LICENSE file.
 
-import '../entrypoint.dart';
 import 'dart:async';
-import 'dart:io';
 
 import 'package:analyzer/analyzer.dart';
-import 'package:pub/src/solver/version_solver.dart';
+import 'package:path/path.dart' as path;
+import 'package:pub/src/dart.dart';
+import 'package:pub/src/entrypoint.dart';
+import 'package:pub/src/io.dart';
 import 'package:pub/src/validator.dart';
+import 'package:source_span/source_span.dart';
 
-/// Returns a map of files -> package names imported or exported within [files].
-Map<String, Iterable<String>> _findUsedPackages(Iterable<String> files) {
-  var packageNames = <String, Iterable<String>>{};
-  for (var file in files) {
-    var usedPackages = <String>[];
-    var compilationUnit = parseDirectives(new File(file).readAsStringSync());
-    for (final directive in compilationUnit.directives) {
-      if (directive is UriBasedDirective) {
-        usedPackages
-            .add(Uri.parse(directive.uri.stringValue).pathSegments.first);
+class StrictDependenciesValidator extends Validator {
+  static bool _isDartFile(String file) => path.extension(file) == '.dart';
+
+  StrictDependenciesValidator(Entrypoint entrypoint) : super(entrypoint);
+
+  /// Returns all pub packages imported or exported within [files].
+  List<_DependencyUse> _findPackages(Iterable<String> files) {
+    var allUses = <_DependencyUse>[];
+    for (var file in files) {
+      var dependencies = new List<_DependencyUse>();
+      try {
+        var contents = readTextFile(file);
+        var directives = parseImportsAndExports(contents, name: file);
+        for (var directive in directives) {
+          var usage = new _DependencyUse(directive, file, contents);
+          if (usage.isPubPackage) {
+            dependencies.add(usage);
+          }
+        }
+        allUses.addAll(dependencies);
+      } on AnalyzerErrorGroup catch (e) {
+        warnings.add('Failed parsing "$file": $e');
```

Comment:
I think for now we should just ignore files that don't parse. I don't want to make users think that pub will verify that their files parse when in fact it's only looking at the first few lines.